src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+    const gs1CompanyPrefix = uri.substring(18, partition);
+    const separator = uri.lastIndexOf('.');
+    const cpref = uri.substring(partition + 1, separator);
+    const rawCPI = gs1CompanyPrefix + cpref;
+    return (`https://id.gs1.org/8010/${webUriPercentEncoder(rawCPI)}`);
+  }
+
+  if (uri.match('^urn:epc:idpat:itip:(([0-9]{6}.[0-9]{7})|([0-9]{7}.[0-9]{6})|([0-9]{8}.[0-9]{5})|([0-9]{9}.[0-9]{4})|([0-9]{10}.[0-9]{3})|([0-9]{11}.[0-9]{2})|([0-9]{12}.[0-9]{1})).[0-9]{2}.[0-9]{2}.\*$')) {
+    const gs1CompanyPrefix = uri.substring(19, partition);
+    const itemRef = uri.substring(partition + 1, (partition + 1 + (13 - gs1CompanyPrefix.length)));
+    const rawGTIN = itemRef.substring(0, 1) + gs1CompanyPrefix + itemRef.substring(1);
+    const piece = uri.substring(34, 36);
+    const total = uri.substring(37, 39);
+    return (`https://id.gs1.org/8006/${addCheckDigitAndZeroPad(rawGTIN, throwError)}${piece}${total}`);
+  }
+
+  if (uri.match('^urn:epc:idpat:upui:(([0-9]{6}.[0-9]{7})|([0-9]{7}.[0-9]{6})|([0-9]{8}.[0-9]{5})|([0-9]{9}.[0-9]{4})|([0-9]{10}.[0-9]{3})|([0-9]{11}.[0-9]{2})|([0-9]{12}.[0-9]{1})).\*$')) {
+    const gs1CompanyPrefix = uri.substring(19, partition);
+    const itemRef = uri.substring(partition + 1, (partition + 1 + (13 - gs1CompanyPrefix.length)));
+    const rawGTIN = itemRef.substring(0, 1) + gs1CompanyPrefix + itemRef.substring(1);
+    return (`https://id.gs1.org/01/${addCheckDigitAndZeroPad(rawGTIN, throwError)}`);
+  }
+
+  if (!uri.match('^https?:(\/\/((([^\/?#]*)@)?([^\/?#:]*)(:([^\/?#]*))?))?((([^?#]*)(\/(01|gtin|8006|itip|8010|cpid|414|gln|417|party|8017|gsrnp|8018|gsrn|255|gcn|00|sscc|253|gdti|401|ginc|402|gsin|8003|grai|8004|giai)\/)([0-9]{4}[^\/]+)?[\/]?(([^?\n]*))?(#([^]*))?)|(\/[A-Za-z_-]{10}$))')) {
+    return uri;
+  }
+
+  if (uri.includes('?')) {
+    uri = uri.substring(0, uri.indexOf('?'));
+  }
+
+  uri = uri.split('/gtin/').join('/01/')
+    .split('/itip/').join('/8006/')
+    .split('/cpid/')
+    .join('/8010/')
+    .split('/gln/')
+    .join('/414/')
+    .split('/party/')
+    .join('/417/')
+    .split('/gsrnp/')
+    .join('/8017/')
+    .split('/gsrn/')
+    .join('/8018/')
+    .split('/gcn/')
+    .join('/255/')
+    .split('/sscc/')
+    .join('/00/')
+    .split('/gdti/')
+    .join('/253/')
+    .split('/ginc/')
+    .join('/401/')
+    .split('/gsin/')
+    .join('/402/')
+    .split('/grai/')
+    .join('/8003/')
+    .split('/giai/')
+    .join('/8004/')
+    .split('/cpv/')
+    .join('/22/')
+    .split('/lot/')
+    .join('/10/')
+    .split('/ser/')
+    .join('/21/');
+
+  if (!uri.match('^https:\/\/id.gs1.org\/(01|8006|8010|414|417|8017|8018|255|00|253|401|402|8003|8004)\/([0-9]{4}[^\/]+)(\/[^\/]+\/[^\/]+)?[\/]?(([^?\n]*))?(#([^\n]*))?|(\/[A-Za-z_-]{10}$)')) {
+    if (uri.indexOf('/00/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/00/'))}`;
+    else if (uri.indexOf('/01/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/01/'))}`;
```

Comment:
`String.includes` can make this more readable. Is there some way to avoid such a large section of `else if`? Extracting a function that uses a map?

Reply:
I'm not sure to understand what you mean. Could you provide an example please?